src/test/java/rx/observables/AsyncOnSubscribeTest.java

```diff
@@ -0,0 +1,408 @@
+package rx.observables;
+
+import static org.junit.Assert.assertEquals;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import rx.Observable;
+import rx.Observable.OnSubscribe;
+import rx.Observer;
+import rx.Subscription;
+import rx.exceptions.TestException;
+import rx.functions.Action0;
+import rx.functions.Action1;
+import rx.functions.Action2;
+import rx.functions.Func0;
+import rx.functions.Func3;
+import rx.internal.util.RxRingBuffer;
+import rx.observers.TestSubscriber;
+import rx.schedulers.TestScheduler;
+
+@RunWith(MockitoJUnitRunner.class)
+public class AsyncOnSubscribeTest {
+
+    @Mock
+    public Observer<Integer> o;
+
+    TestSubscriber<Integer> subscriber;
+    
+    @Before
+    public void setup() {
+        subscriber = new TestSubscriber<Integer>(o);
+    }
+    
+    @Test
+    public void testSerializesConcurrentObservables() throws InterruptedException {
+        final TestScheduler scheduler = new TestScheduler();
+        OnSubscribe<Integer> os = AsyncOnSubscribe.createStateful(new Func0<Integer>(){
+            @Override
+            public Integer call() {
+                return 1;
+            }}, 
+        new Func3<Integer, Long, Observer<Observable<? extends Integer>>, Integer>(){
+            @Override
+            public Integer call(Integer state, Long requested, Observer<Observable<? extends Integer>> observer) {
+                if (state == 1) {
+                    Observable<Integer> o1 = Observable.just(1, 2, 3, 4)
+                            .delay(100, TimeUnit.MILLISECONDS, scheduler);
+                    observer.onNext(o1);
+                }
+                else if (state == 2) {
+                    Observable<Integer> o = Observable.just(5, 6, 7, 8);
+                    observer.onNext(o);
+                }
+                else 
+                    observer.onCompleted();
+                return state + 1;
+            }});
+        // initial request emits [[1, 2, 3, 4]] on delay
+        Observable.create(os).subscribe(subscriber);
+        // next request emits [[5, 6, 7, 8]] firing immediately
+        subscriber.requestMore(2); 
```

Comment:
I believe `TestSubscriber` requests `Long.MAX_VALUE` by default so these requests do nothing. In addition, I think if you had a state 3 with yet another observable, a sequence of requests of 1 would overflow the `concat` queue and yield a `MissingBackpressureException` - however, I can't check this scenario right now.